### DISCUSSION TITLE
GoogleTest Tests not found

### DISCUSSION THREAD

**Discussion by tridivb** [cpp-for-yourself/lectures-and-homeworks#43] [Homeworks Q&A] [answered]
Hi,

Are we supposed to link the googletest subdirectory somehow other than using `add_subdirectory` in the root CmakeLists.txt? Basically, I downloaded the googletest framework manually, unzipped it into the `externals/googletest` subdirectory. While trying to run the tests, despite successful compilation , the framework just says that it cannot find any tests.

![image](https://user-images.githubusercontent.com/28273134/209627243-361229a9-9a62-4276-9448-592c0b26b242.png)

**Comment by niosus** [comment, marked as answer]
Sounds like you have to actually add some c++ code, add it as an executable to CMake and register as a test. Did you do that?

**Reply by tridivb** [reply]
Haha! Yes, that would be the obvious issue except that I basically followed your guidelines including the code other than changing the name of the project in all the relevant places. I did find an old stackoverflow post which mentions that defining a main function is necessary for this but I am guessing this is obsolete as even the googletest sample codes don't include it in their repo and the post itself is very old.

https://stackoverflow.com/questions/50222112/google-test-not-finding-any-tests

**Reply by niosus** [reply]
Interesting. Could you link me to your project? Maybe I'll spot what is going on there. Googletest provides the main function the way I described the setup in the video.

**Reply by tridivb** [reply]
Sure! I have created a temporary repo and added you to it.

**Reply by niosus** [reply]
Hmm, strange. I cloned your project, and everything builds and tests as expected with the following commands:
```cmd
cmake -S . -B build
cmake --build build
GTEST_COLOR=1 ctest --test-dir build --output-on-failure
```
So it is very strange that it doesn't work for you. Let's try to find out what is different about your setup in comparison to mine. I have a feeling that it's because of these flags: `--test-dir build`. These options won't work for an older version of `ctest`. To verify that this is the cause, you could try to run the same command from the build folder:
```cmd
cd build
GTEST_COLOR=1 ctest --output-on-failure
``` 
Hope this helps.

Also, here are some things I've noticed in your project that are different from what I was talking about in the video and created a PR into your repo with the fixes 😉

**Reply by tridivb** [reply]
Voila! It did work from inside the `build` directory. I am guessing you are using cmake 3.20+. Mine is 3.16.3. Based on the following post, I need to include `add_test` to the cmake file somewhere to make it work without having to run `ctest` from inside build everytime (unless I just update the cmake).

https://stackoverflow.com/questions/43981966/cmake-how-to-specify-directory-where-ctest-should-look-for-executables

**Reply by niosus** [reply]
Glad it worked out! 🙏
